src/cpu/flexcpu/inflight_inst.hh

```diff
@@ -0,0 +1,592 @@
+        // InflightInst object will apply to committed state completed?
+        Tick completionTick;
+        // When was this InflightInst finally applied to committed state?
+        Tick commitTick;
+        // If squashed, when was this InflightInst squashed?
+        Tick squashTick;
+    };
+
+    struct DataSource
+    {
+        std::weak_ptr<InflightInst> producer;
+        int8_t resultIdx;
+    };
+
+    class MemIface
+    {
+      public:
+        virtual Fault readMem(std::shared_ptr<InflightInst> inst, Addr addr,
+                              uint8_t *data, unsigned int size,
+                              Request::Flags flags,
+                              const std::vector<bool>& byte_enable)
+        { panic("MemIface::readMem() not implemented!"); }
+
+        virtual Fault initiateMemRead(std::shared_ptr<InflightInst> inst,
+                                      Addr addr, unsigned int size,
+                                      Request::Flags flags,
+                                      const std::vector<bool>& byte_enable)
+        { panic("MemIface::initiateMemRead() not implemented!"); }
+
+        virtual Fault writeMem(std::shared_ptr<InflightInst> inst,
+                               uint8_t *data, unsigned int size, Addr addr,
+                               Request::Flags flags, uint64_t *res,
+                               const std::vector<bool>& byte_enable)
+        { panic("MemIface::writeMem() not implemented!"); }
+
+        static const MemIface unimplemented;
+    };
+
+    // Maybe another interface for more nuanced
+
+  protected:
+    ThreadContext* backingContext;
+    TheISA::ISA* backingISA;
+    MemIface* backingMemoryInterface;
+
+    // Where am I in completing this instruction?
+    Status _status;
+    bool _squashed = false;
+
+    // What is this instruction I'm executing?
+    InstSeqNum _seqNum;
+    InstSeqNum _issueSeqNum;
+    StaticInstPtr instRef;
+
+    /**
+     * A storage for the exact PCState at the time of executing this
+     * instruction. Any changes to the PC through the instruction can also be
+     * captured in this variable, allowing this variable also to serve as the
+     * result storage for changes to the PCState. If this variable is changed
+     * in this fashion however, we may need to flush speculative state.
+     */
+    PCState _pcState;
+
+    Trace::InstRecord* _traceData = nullptr;
+
+    TimingRecord _timingRecord;
+
+    // Callbacks made during a state transition.
+    std::vector<std::function<void()>> commitCallbacks;
+    std::vector<std::function<void()>> completionCallbacks;
+
+    std::vector<std::function<void()>> effAddrCalculatedCallbacks;
+
+    // How many dependencies have to be resolved before I can execute?
+    size_t remainingDependencies = 0;
+
+    // How many dependencies have to be resolved before I can go to memory?
+    size_t remainingMemDependencies = 0;
+
+    std::vector<std::function<void()>> readyCallbacks;
```

Comment:
Group all callbacks together?

Reply:
I prefer to keep xReady and remainingXDependencies together though because they're functionally closely related. All the callbacks are otherwise in those 10 or so lines.